File: src/python/oop/VendingMachine/vending/vending_machine.py

from .states import *
from .inventory_manager import *
from .product import *

class VendingMachine:
    def __init__(self):
        self._inventory_manager = InventoryManager()
        self._state = IdleState(self)
        self._state.process()

    def add_product(self, rack_number: str, product: Product, count: int) -> None:
        if self._inventory_manager.add_product_to_rack(rack_number, product, count):
            print(f"Added {product.name} to the {rack_number} rack")
        else:
            print(f"Failed add {product.name} to the {rack_number} rack")

    def insert_money(self, amount: float) -> None:
        if not isinstance(self._state, WaitingForMoneyState):
            return

        self._state.context.balance += amount
        self._state.process()

    def select_product(self, code: str) -> None:
        if not isinstance(self._state, SelectProductState):
            return

        self._state.context.rack_code = code
        self._state.process()

    def pickup_change(self) -> None:
        if not isinstance(self._state, ReturnChangeState):
            return

        self._state.process()

    def dispense_product(self) -> None:
        if not isinstance(self._state, DispenseProductState):
            return

        self._state.process()

    def _notify(self, message: str) -> None:
        print(message)

File: src/csharp/oop/VendingMachine/Vending/VendingMachine.cs
using VendingMachine.Vending.States;

namespace VendingMachine.Vending;

public sealed class VendingMachine
{
    private OrderStateBase _state;

    public VendingMachine(InventoryManager inventoryManager)
    {
        InventoryManager = inventoryManager;
        _state = new IdleState(this);
        _state.Process();
    }

    public InventoryManager InventoryManager { get; }

    public void AddProduct(string rackNumber, Product product, int count)
    {
        if (InventoryManager.AddProduct(rackNumber, product, count))
        {
            Notify($"Added {product.Name} to the {rackNumber} rack");
        }
        else
        {
            Notify($"Failed add {product.Name} to the {rackNumber} rack");
        }
    }

    public void InsertMoney(decimal amount)
    {
        if (_state is not WaitingForMoneyState)
        {
            return;
        }

        _state.Context.Balance = decimal.Add(_state.Context.Balance ?? decimal.Zero, amount);
        _state.Process();
    }

    public void SelectProduct(string code)
    {
        if (_state is not SelectProductState)
        {
            return;
        }

        _state.Context.RackNumber = code;
        _state.Process();
    }

    public void PickupChange()
    {
        if (_state is not ReturnChangeState)
        {
            return;
        }

        _state.Process();
    }

    public void DispenseProduct()
    {
        if (_state is not DispenseProductState)
        {
            return;
        }

        _state.Process();
    }

    internal void Notify(string message)
        => Console.WriteLine(message);

    internal void SetState(OrderStateBase state)
        => _state = state ?? throw new ArgumentNullException(nameof(state));
}
